### PR TITLE
Ane 1117

### PR DESCRIPTION
**Pull request by csasarak**
# Overview

This adds support for comment and code normalization in C.

## Acceptance criteria

* It should be able to normalize C code by removing all comments (comment normalization).
* It should be able to normalize C code by removing all comments and turning all pieces of contiguous whitespace into single spaces (code normalization).

## Testing plan

I used automated test cases.

## Risks

None in particular.

## References

- _[ANE-1117](https://fossa.atlassian.net/browse/ANE-1117): https://fossa.atlassian.net/browse/ANE-1117

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

[ANE-1117]: https://fossa.atlassian.net/browse/ANE-1117?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: snippets/src/language/snippet_context.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    /// Get content from the snippet which is not in ranges covered by the provided nodes.
+    pub fn retrieve_negative_content(
+        &self,
+        nodes: impl Iterator<Item = &'a Node<'a>>,
+    ) -> impl Iterator<Item = &'a [u8]> {
+        let mut start_byte = 0;
+        let mut slices = Vec::new();
+
+        // Find every non-comment section of text from the original content into a sequence
```
Comment (spatten) [inline snippets/src/language/snippet_context.rs]: [nit] This function is more general than removing non-comment nodes, so the comment here should reflect that

File: snippets/src/language/snippet_context.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        SnippetContext {
+            offset,
+            context_nodes,
+            location,
+            content,
+        }
+    }
+
+    /// Get content from the snippet which is not in ranges covered by the provided nodes.
+    pub fn retrieve_negative_content(
```
Comment (spatten) [inline snippets/src/language/snippet_context.rs]: I'm not 100% sure about this name

What this function is doing is removing the provided nodes

Maybe `without_provided_nodes` or `remove_nodes`?

File: snippets/src/language/normalize_code.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+use super::{normalize_comments, snippet_context::SnippetContext};
+use crate::text::normalize_space;
+
+/// Perform both comment and code normalization.
+///
+/// Uses [`super::normalize_comments`] and [`crate::text::normalize_space`].
+#[tracing::instrument(skip_all)]
+pub fn normalize_code<'a>(context: &'a SnippetContext) -> Cow<'a, [u8]> {
+    let comments = normalize_comments(context);
```
Comment (spatten) [inline snippets/src/language/normalize_code.rs]: [nit] [pedantic] This variable should be `without_comments` or `uncommented` or something like that

**Review by spatten — APPROVED** [review]
Looks good to me!

File: snippets/src/language/snippet_context.rs (hunk 2 above)
Comment (csasarak) [reply snippets/src/language/snippet_context.rs]: I think those names might imply that we are altering the `SnippetContext` itself which is not correct. I think your point around the name being confusing is a good one. I'll have to think about it.

I think that it's generally good to have a structure that represents this concept, but I am still a little uncomfortable with the interface I've defined for it. I think ideally, this function would return a whole new `SnippetContext` with a new view into the modified context. Then the interface is effectively `context.filter_nodes(|f| <some predicate>)`. 

Initially I was thinking it'd be hard to keep track of all the offsets as nodes are added and removed. But actually if someone ever tries to view the content I can just generate it on-demand from the original. Maybe this is more possible than I thought it was.

Comment (csasarak) [reply snippets/src/language/snippet_context.rs]: I spent a little time with that above and found that it doesn't work so well. `Node`s can overlap, so filtering nodes and outputting content from the remaining ones doesn't do what you'd expect. For example:
```
int main () {
// I am a comment
}
```
Has a node of kind comment which contains the comment text. But it also has a node called "translation_unit" that covers the entire parsed function, comment and all. So simply removing the comments and copying the remaining text out doesn't work in a straightforward way. I think this is probably why I did such a weird solution to begin with.

For now, I've just changed the name to `retrieve_content_around_nodes`.

File: snippets/src/language/normalize_code.rs (hunk 3 above)
Comment (csasarak) [reply snippets/src/language/normalize_code.rs]: I turns out I don't need a variable at all, so I've removed it.

File: snippets/src/language/normalize_comments.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+use crate::tree_sitter_consts::NODE_KIND_COMMENT;
+use std::borrow::Cow;
+
+use super::snippet_context::SnippetContext;
+
+/// Remove all comment node text from the given content.
+/// In general, this function should work in any language that produces "comment" type nodes from tree_sitter.
+#[tracing::instrument(skip_all)]
```
Comment (jssblck) [inline snippets/src/language/normalize_comments.rs]: [nit] Prefer to name the literal. e.g.:
```suggestion
/// In general, this function should work in any language for which treesitter produces nodes whose `kind` equals [`NODE_KIND_COMMENT`].
#[tracing::instrument(skip_all)]
```

**Pull request merged**